Autogenerate flint headers

videlec opened the pull request:
We
- write down a script to auto-generate flint header files (currently at https://github.com/videlec/flint-header-sage-autogen)
- replace most of the files in `src/sage/libs/flint` by auto-generated files. Doing so, the custom sage code in `src/sage/libs/flint` is always contained in files suffixed by `_sage.pxd`/`_sage.pyx`.
- deprecate the files in `src/sage/libs/arb/`

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

#35848 

#### Upstream issues and patches

- https://github.com/flintlib/flint/pull/1523
- https://github.com/flintlib/flint/issues/1529
- https://github.com/flintlib/flint/pull/1532
- https://github.com/flintlib/flint/issues/1535
- https://github.com/flintlib/flint/pull/1536
- https://github.com/cython/cython/issues/5779
- https://github.com/flintlib/flint/issues/1653
- https://github.com/flintlib/flint/pull/1655

videlec committed:
automatic generation of flint headers

videlec committed:
two fixes

videlec committed:
do not include fft_small.h

videlec committed:
do not include machine_vectors.h

mkoeppe commented on build/pkgs/sagelib/dependencies:
```diff
@@ -1,4 +1,4 @@
-FORCE $(SCRIPTS) arb boost_cropped $(BLAS) brial cliquer cypari cysignals cython ecl eclib ecm flint libgd gap giac givaro glpk gmpy2 gsl iml importlib_metadata importlib_resources jupyter_core lcalc lrcalc_python libbraiding libhomfly libpng linbox m4ri m4rie memory_allocator mpc mpfi mpfr $(MP_LIBRARY) ntl numpy pari pip pkgconfig planarity ppl pplpy primesieve primecount primecountpy pycygwin  ...
+FORCE $(SCRIPTS) boost_cropped $(BLAS) brial cliquer cypari cysignals cython ecl eclib ecm flint libgd gap giac givaro glpk gmpy2 gsl iml importlib_metadata importlib_resources jinja2 jupyter_core lcalc lrcalc_python libbraiding libhomfly libpng linbox m4ri m4rie memory_allocator mpc mpfi mpfr $(MP_LIBRARY) ntl numpy pari pip pkgconfig planarity ppl pplpy primesieve primecount primecountpy pycygw ...
```
careful, this is undoing changes done elsewhere

videlec replied:
I do not think that concerns this #36449, does it? You might want to have a look at #35848.

mkoeppe replied:
You are right, it's coming from there

videlec replied:
Thanks for having a look.

mezzarobba replied:
> careful, this is undoing changes done elsewhere

This looks like a bad rebase. Thanks!

mkoeppe commented on src/sage/schemes/plane_conics/con_rational_function_field.py:
```diff
@@ -130,19 +130,16 @@ def has_rational_point(self, point=False, algorithm='default',
             (True, (-3 : (t + 1)/t : 1))
-
-            sage: R.<t> = FiniteField(23)[]
-            sage: C = Conic([2, t^2 + 1, t^2 + 5])
```
the `# optional` is not needed here.

videlec replied:
@mezzarobba also coming from at #35848

mkoeppe commented on src/sage/schemes/plane_conics/con_rational_function_field.py:
```diff
@@ -130,19 +130,16 @@ def has_rational_point(self, point=False, algorithm='default',
-
-            sage: # needs sage.rings.number_field
-            sage: F.<i> = QuadraticField(-1)
-            sage: R.<t> = F[]
```
the version with the block `# needs` is the correct one

videlec replied:
@mezzarobba also coming from at #35848

videlec committed:
more autogenerated headers

videlec committed:
modify arb headers

videlec committed:
more flint declarations

videlec committed:
fix arb headers

videlec committed:
more autogenerated flint headers

videlec committed:
adaptation of sage source code

videlec committed:
make linter happier

videlec committed:
number_of_partitions moved

videlec committed:
fix doctests in libs/flint/

videlec committed:
int -> bint

videlec committed:
fix change in imports

videlec committed:
Flint documentation

videlec committed:
deprecations

videlec committed:
fix declarations and test inclusion of all headers in flint_sage.pyx

videlec committed:
some fixes in flint headers

videlec committed:
remove extra parenthesis to make cython less confused

videlec committed:
noexcept

videlec committed:
clean fmpq_poly_sage[.pxd,.pyx]

videlec committed:
safer flint_wrap.h

videlec committed:
fix import of n_factor_to_list in integer.pyx

videlec committed:
fix import of qsieve in integer.pyx

mezzarobba commented on src/sage/libs/flint/acb_macros.pxd:
```diff
@@ -0,0 +1,11 @@
+# Macros from acb.h
+# See https://github.com/flintlib/flint/issues/152
```
wrong issue number? (should be 1529, I think)

videlec replied:
fixed by you!

mezzarobba commented on src/sage/libs/flint/fmpz_factor_extra.pyx:
shouldn't this module be called `fmpz_factor_sage` for consistency?

videlec replied:
done

mezzarobba commented on src/sage/libs/flint/fmpz.pxd:
```diff
@@ -1,195 +1,1126 @@
+
+    __mpz_struct * COEFF_TO_PTR(fmpz f) noexcept
+
+    int COEFF_IS_MPZ(fmpz f) noexcept
```
conflicts with previous declaration in `types.pxd`

videlec replied:
Indeed.`COEFF_IS_MPZ` is a macro in `flint/fmpz.h`. However, in `flint/doc/source/fmpz.rst` it is declared as a function
```rest
.. function:: int COEFF_IS_MPZ(fmpz f)
```
and hence automatically declared by the autogeneration script. It ought to be a `.. macro` in `flint/doc/source/fmpz.rst` and be in `fmpz_macros.pxd` rather than `types.pxd`.

mezzarobba reviewed (commented):
Hi Vincent,

This looks great! See the other comments and https://github.com/videlec/flint-header-sage-autogen/pull/4 for the few minor issues I noticed.

Two other remarks, not necessarily for this ticket:
* I like the fact that your scripts can copy the documentation to the `pxd` files, but I am not sure if it is a good idea to do it in the version shipped with sage;
* what about moving the scripts themselves to `src/sage_setup/autogen` (which already contains scripts with a similar status, afaict)?

videlec committed:
include autogeneration files

videlec committed:
more types

videlec committed:
sort pxd_list

videlec committed:
regenerate flint cython headers

videlec committed:
clean macros

videlec committed:
remove COEFF_IS_MPZ from types.pxd

videlec committed:
do generate flint.pxd

videlec committed:
some more notes in README

videlec committed:
fix fmpz_factor_sage import

videlec committed:
remove specialized acb_poly_revert_series that disappeared from flint

videlec committed:
linting

mezzarobba commented on src/sage_setup/autogen/flint/README.md:
```diff
@@ -0,0 +1,25 @@
+
+4. Set the environment variable `FLINT_GIT_DIR`
+
+5. Run the `run.py` script, eg `python autogen.py`
```
`../flint_autogen.py`?

videlec replied:
indeed

videlec committed:
fix README

mezzarobba reviewed (approved):
thanks!

videlec committed:
make sage.libs.flint.*.py a cython modules again

Otherwise, a non cleaned installation would still use the "old" cython
version rather than the python one.

videlec committed:
fix import in combinat.py

videlec committed:
fix sage_flint feature

videlec committed:
fix import of dedekind_sum

videlec committed:
include mpfr in flint_wrap
